### PR TITLE
[PW-2029] Configuration to charge orders based on storeview or global currency

### PR DESCRIPTION
**Pull request by acampos1916**
**Description**
The plugin was designed to charge the order amount using the storeview or `display` currency. We're introducing a new configuration to also select the Global/Website or `base` currency to process payments. `display` remains as the default option.

A new column has been introduced for the sales_order table to store the order's currency based on this new configuration at the time of placement, with a new event observer to populate it. This has been done so refunds can be processed with the correct currency.

**Tested scenarios**
CaptureDataBuilder
CaptureDataBuilder Openinvoice
CheckoutDataBuilder Openinvoice
Cron _addStatusHistoryComment
Cron _isTotalAmount
Cron _prepareInvoice
Cron _refundOrder
Cron _setPaymentAuthorized
GooglePayConfigProvider
PayByMail payment
Payment methods call
PaymentDataBuilder
RefundDataBuilder

**To test**
Currency exchange settings
Matching currencies
TerminalApi

**Fixed issue**:  PW-2029

**TODO**
Docs on new wrapper usage

### CODE REVIEW

File: Setup/UpgradeSchema.php [hunk 1]
```diff
@@ -402,4 +406,33 @@ public function updateSchemaVersion540(SchemaSetupInterface $setup)
...
+     *
+     * New sales_order column for the currency charged based on the Adyen config option (base or display)
+     *
+     * @param SchemaSetupInterface $setup
+     * @return void
+     */
+    public function updateSchemaVersion670(SchemaSetupInterface $setup)
+    {
+        $connection = $setup->getConnection();
+        $tableName = $setup->getTable('sales_order');
```
Comment (acampos1916) [inline Setup/UpgradeSchema.php:421]: Check if it works with prefixes

File: Observer/AdyenSalesOrderChargedCurrencyObserver.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Observer;
+
+use Adyen\Payment\Helper\ChargedCurrency;
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\Payment\Observer\AbstractDataAssignObserver;
+use Magento\Quote\Api\Data\PaymentInterface;
```
Comment (acampos1916) [inline Observer/AdyenSalesOrderChargedCurrencyObserver.php:31]: Remove unused deps

File: Observer/AdyenSalesOrderChargedCurrencyObserver.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    ) {
+        $this->chargedCurrency = $chargedCurrency;
+    }
+
+    public function execute(Observer $observer)
+    {
+        /** @var \Magento\Sales\Model\Order $order */
+        $order = $observer->getEvent()->getOrder();
+
+        //TODO only save it for Adyen PMs?
```
Comment (acampos1916) [inline Observer/AdyenSalesOrderChargedCurrencyObserver.php:52]: How to?

File: Helper/Config.php [hunk 4]
```diff
@@ -153,6 +154,17 @@ public function isStoreAlternativePaymentMethodEnabled($storeId = null)
...
 
+    /**
+     * Retrive charged currency selection (base or display)
+     *
+     * @param null|int|string $storeId
+     * @return mixed
+     */
+    public function getChargedCurrency($storeId = null)
+    {
+        return $this->adyenHelper->getAdyenAbstractConfigData(self::XML_CHARGED_CURRENCY, $storeId);
```
Comment (acampos1916) [inline Helper/Config.php:165]: Is it worth it to set base as default and change to display on upgrade?

File: Gateway/Command/PayByMailCommand.php [hunk 5]
```diff
@@ -143,10 +149,17 @@ protected function getFormFields($payment, $paymentAmount = false)
             $hmacKey = $this->_adyenHelper->getHmacPayByMail();
         }
 
-        $amount = $this->_adyenHelper->formatAmount(
-            $paymentAmount ?: $order->getGrandTotal(),
-            $orderCurrencyCode
-        );
+        if ($paymentAmount) {
```
Comment (acampos1916) [inline Gateway/Command/PayByMailCommand.php:152]: Check new helper usage

File: Gateway/Request/CaptureDataBuilder.php [hunk 6]
```diff
@@ -55,13 +65,15 @@ public function build(array $buildSubject)
     {
         /** @var \Magento\Payment\Gateway\Data\PaymentDataObject $paymentDataObject */
         $paymentDataObject = \Magento\Payment\Gateway\Helper\SubjectReader::readPayment($buildSubject);
-        $amount = \Magento\Payment\Gateway\Helper\SubjectReader::readAmount($buildSubject);
+        $amount = \Magento\Payment\Gateway\Helper\SubjectReader::readAmount($buildSubject); //TODO which one of these to use?
```
Comment (acampos1916) [inline Gateway/Request/CaptureDataBuilder.php:68]: Use helper

File: etc/module.xml [hunk 7]
```diff
@@ -24,7 +24,7 @@
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
 
-    <module name="Adyen_Payment" setup_version="5.4.0">
+    <module name="Adyen_Payment" setup_version="6.7.0">
```
Comment (acampos1916) [inline etc/module.xml:27]:
```suggestion
    <module name="Adyen_Payment" setup_version="7.0.0">
```

File: etc/adminhtml/system/adyen_required_settings.xml [hunk 8]
```diff
@@ -69,7 +69,12 @@
             <comment><![CDATA[Provide the unique live url prefix: <strong>[random]-[company name]</strong> from the "API URLs and Response" menu in the Adyen Customer Area. For more information, please check <a href="https://docs.adyen.com/developers/development-resources/live-endpoints#checkoutendpoints">  our documentation</a>.]]></comment>
             <config_path>payment/adyen_abstract/live_endpoint_url_prefix</config_path>
         </field>
-
+        <field id="charged_currency" translate="label" type="select" sortOrder="80" showInDefault="1" showInWebsite="0" showInStore="0">
+            <label>Charged currency</label>
+            <comment><![CDATA[Currency used for Adyen payment processing.]]></comment>
+            <source_model>Adyen\Payment\Model\Config\Source\ChargedCurrency</source_model>
+            <config_path>payment/adyen_abstract/charged_currency</config_path>
+        </field>
```
Comment (acampos1916) [inline etc/adminhtml/system/adyen_required_settings.xml:77]: Hide this field

File: Model/Cron.php [hunk 9]
```diff
@@ -683,6 +703,20 @@ protected function _declareVariables($notification)
...
+    }
+
+    /**
+     * Declare private variables with order charged amount and currency
+     *
+     * @return void
+     */
+    private function declareOrderVariables()
+    {
+        $orderAmountCurrency = $this->chargedCurrency->getOrderAmountCurrency($this->_order, false);
```
Comment (acampos1916) [inline Model/Cron.php:717]: New param name, useStoredConfig

File: Model/Cron.php [hunk 10]
```diff
@@ -1394,8 +1425,7 @@ protected function _refundOrder()
             // refund is done through adyen backoffice so create a credit memo
             $order = $this->_order;
             if ($order->canCreditmemo()) {
-                $currency = $this->_order->getOrderCurrencyCode();
-                $amount = $this->_adyenHelper->originalAmount($this->_value, $currency);
+                $amount = $this->_adyenHelper->originalAmount($this->_value, $this->orderCurrency);
```
Comment (acampos1916) [inline Model/Cron.php:1428]: Use notification currency

File: Model/Cron.php [hunk 11]
```diff
@@ -1831,7 +1860,10 @@ protected function _isTotalAmount($paymentId, $orderCurrencyCode)
         );
 
         // get total amount of the order
-        $grandTotal = (int)$this->_adyenHelper->formatAmount($this->_order->getGrandTotal(), $orderCurrencyCode);
+        $grandTotal = (int)$this->_adyenHelper->formatAmount(
+            $this->orderAmount,
+            $orderCurrencyCode
```
Comment (acampos1916) [inline Model/Cron.php:1865]: Update with $this->orderCurrency

File: Model/Cron.php [hunk 12]
```diff
@@ -1569,8 +1599,7 @@ protected function _prepareInvoice()
         }
 
         // validate if amount is total amount
-        $orderCurrencyCode = $this->_order->getOrderCurrencyCode();
-        $amount = $this->_adyenHelper->originalAmount($this->_value, $orderCurrencyCode);
+        $amount = $this->_adyenHelper->originalAmount($this->_value, $this->orderCurrency);
```
Comment (acampos1916) [inline Model/Cron.php:1602]: Use notification currency

File: Model/Cron.php [hunk 13]
```diff
@@ -1955,14 +1987,13 @@ protected function _setPaymentAuthorized($manualReviewComment = true, $createInv
         $this->_adyenLogger->addAdyenNotificationCronjob('Set order to authorised');
 
         // if full amount is captured create invoice
-        $currency = $this->_order->getOrderCurrencyCode();
         $amount = $this->_value;
-        $orderAmount = (int)$this->_adyenHelper->formatAmount($this->_order->getGrandTotal(), $currency);
+        $formattedOrderAmount = (int)$this->_adyenHelper->formatAmount($this->_order->getGrandTotal(), $this->orderCurrency);
```
Comment (acampos1916) [inline Model/Cron.php:1991]: Update with $this->orderAmount

File: Model/AdyenInitiateTerminalApi.php [hunk 14]
```diff
@@ -162,7 +171,7 @@ public function initiate($payload)
                                 [
                                     'AmountsReq' =>
                                         [
-                                            'Currency' => $quote->getCurrency()->getQuoteCurrencyCode(),
+                                            'Currency' => $this->chargedCurrency->getQuoteAmountCurrency($quote),
                                             'RequestedAmount' => doubleval($quote->getGrandTotal())
```
Comment (acampos1916) [inline Model/AdyenInitiateTerminalApi.php:175]: Use wrapper

File: Gateway/Request/CheckoutDataBuilder.php [hunk 15]
```diff
@@ -283,11 +299,13 @@ protected function getOpenInvoiceData($order): array
         // Shipping cost
         if ($cart->getShippingAddress()->getShippingAmount() > 0 || $cart->getShippingAddress()->getShippingTaxAmount(
             ) > 0) {
-            $priceExcludingTax = $cart->getShippingAddress()->getShippingAmount() - $cart->getShippingAddress(
-                )->getShippingTaxAmount();
+
+            $shippingAmountCurrency=$this->chargedCurrency->getQuoteShippingAmountCurrency($cart);
```
Comment (acampos1916) [inline Gateway/Request/CheckoutDataBuilder.php:303]:
```suggestion
            $shippingAmountCurrency = $this->chargedCurrency->getQuoteShippingAmountCurrency($cart);
```

File: Helper/Config.php (hunk 4 above)
Comment (acampos1916) [reply Helper/Config.php:165]: Can be discussed upon release

File: Model/Cron.php (hunk 9 above)
Comment (acampos1916) [reply Model/Cron.php:717]: useStoredConfig doesn't really reflect the data stored. Can be changed to useStoredScopeConfigOnOrderPlacement but is too long

**Review by cyattilakiss — APPROVED** [review]

**Review by msilvagarcia — APPROVED** [review]

**Pull request merged**